akka-http2-support/src/main/scala/akka/http/impl/engine/http2/Http2StreamHandling.scala

```diff
@@ -74,11 +76,38 @@ private[http2] trait Http2StreamHandling { self: GraphStageLogic with LogHelper
 
   private def updateState(streamId: Int, handle: IncomingStreamState => IncomingStreamState): Unit = {
     val oldState = streamFor(streamId)
-    val newState = handle(oldState)
+    val newState: IncomingStreamState = handle(oldState)
+
+    // only validate maxConcurrentStreams compliance if we'd increment the number
+    // of streams in one of the open states. This way, when the value of SETTINGS_MAX_CONCURRNET_STREAMS
+    // decreases, inflight streams can still finish and close.
+    val needsMaxConcurrentComplianceCheck = (oldState, newState) match {
+      case (Idle, _) => true
+      case _         => false
+    }
+
+    if (needsMaxConcurrentComplianceCheck) {
+      // incomingStreams contains all the open streams and an extra one that's Idle9
+      val currentConcurrentStreams = incomingStreams.size - 1
+      if (!Http2Compliance.compliesWithMaxConcurrentStreams(currentConcurrentStreams, maxConcurrentStreams)) {
+        pushGOAWAY(
```

Comment:
```suggestion
        // 5.1.2 An endpoint that receives a HEADERS frame that causes its advertised 
        // concurrent stream limit to be exceeded MUST treat this as a stream error 
        // (Section 5.4.2) of type PROTOCOL_ERROR or REFUSED_STREAM. The choice of error 
        // code determines whether the endpoint wishes to enable automatic retry 
        // (see Section 8.1.4) for details).
        pushGOAWAY(
```

I chose GOAWAY so the peer can retry.